hkmc2/shared/src/main/scala/hkmc2/codegen/HandlerLowering.scala

```diff
@@ -26,56 +26,75 @@ object HandlerLowering:
     def next = p.selN(nextIdent)
     def last = p.selN(lastIdent)
     def contTrace = p.selN(contTraceIdent)
+    
+  extension (b: Block) def userDefinedVars: Set[Local] = b.definedVars.collect:
+    case s: VarSymbol => s
   
-  private case class LinkState(res: Path, cls: Path, uid: StateId)
+  private case class LinkState(res: Local, cls: Path, uid: StateId)
   
   // isTopLevel:
   // whether the current block is the top level block, as we do not emit code for continuation class on the top level
   // since we cannot return an effect signature on the top level (we are not in a function so return statement are invalid)
-  // and we do not have any `return` statement in the top level block so we do not need the `runtime.Return` workarounds.
   // contName: the name of the continuation class
   // ctorThis: the path to `this` in the constructor, this is used to insert `return this;` at the end of constructor.
   // linkAndHandle:
   // a function that takes a LinkState and returns a block that links the continuation class and handles the effect
   // this is a convenience function which initializes the continuation class in function context or throw an error in top level
   private case class HandlerCtx(
-    isTopLevel: Bool,
-    isHandlerBody: Bool,
-    contName: Str,
-    ctorThis: Option[Path],
-    linkAndHandle: LinkState => Block
-  )
+        isTopLevel: Bool,
+        isHandlerBody: Bool,
+        contName: Str,
+        ctorThis: Option[Path],
+        debugInfo: DebugInfo,
+        linkAndHandle: LinkState => Block,
+    ):
+    def nestDebugScope(locals: Set[Local], localsFn: Path) = copy(debugInfo = DebugInfo(
+      debugInfo.inScopeLocals ++ locals, S(localsFn)))
+  
+  // inScopeLocals: Local variables that are in scope.
+  // prevLocalsFn: The function that gets the outer function's locals.
+  private case class DebugInfo(
+      inScopeLocals: Set[Local],
```

Comment:
These should probably only have one line of indentation (the point of having more indentation is to have them on a different indentation level than the class members and methods).